MH-12815 delete series with events option

doofy opened the pull request:
This PR adds two new features concerning series deletion:

1. Series deletion uses a new request to check if a series contains events and warns about this in the deletion dialog (adminUI delete in series table or bulk delete via start task)
2. SeriesEndpoint now has a config file `etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg` where you can specify if it is allowed to delete a series containing events `series.hasEvents.delete.allow=true`. Admin UI will check for this and prevent deletion. For this a new request was added which passes config values from the config file.

Example usecase: You can configure events to require a series to be set. `etc/org.opencastproject.ui.metadata.CatalogUIAdapterFactory-episode-common.cfg: property.series.required=true`. Setting this will not prevent the user from deleting the series.

Here some example screenshots:

**Delete a single series with no events**

![screen shot 2018-05-14 at 17 13 17](https://user-images.githubusercontent.com/1590263/40006398-97fbf536-579a-11e8-8f42-4a97bb89cafd.png)

**Delete a single series with events (default configuration, allowed)**

![screen shot 2018-05-14 at 17 13 31](https://user-images.githubusercontent.com/1590263/40006442-b6f8f1f0-579a-11e8-8fdb-1edfe01fa3e3.png)

The assumption here is that a user would normally not want to delete a series if it contains events - even if she or he is allowed to.

**Delete a single series with events (disallowed by configuration)**

![screen shot 2018-05-14 at 17 16 03](https://user-images.githubusercontent.com/1590263/40006474-c7347f9e-579a-11e8-8ffe-66293e7c7fd9.png)

Admittedly not the best usability - but if the user deletes the series, this can cause problems if Opencast clients rely on it.

**Delete multiple series with or without events (default configuration, allowed)**

![screen shot 2018-05-14 at 17 19 22](https://user-images.githubusercontent.com/1590263/40006574-0628a64e-579b-11e8-8c25-9ae63ef83d09.png)

**Delete multiple series with events (disallowed by configuration)**

![screen shot 2018-05-14 at 17 18 22](https://user-images.githubusercontent.com/1590263/40006593-137d400c-579b-11e8-951a-6d3923435d0a.png)

Background information:

In our environment, Series are mandatory. Events not belonging to any Series will cause problems in surrounding systems (learning management plugins, video portal) that expect Events to belong to Series.
This PR addresses this issue for the most prominent case: A user deletes a Series in the Admin UI - leaving orphaned events which - in our case - present inconsistent data.

A solution we would have preferred was to delete all events when deleting a series. That is, however, out of scope.

The usability could certainly be improved (presenting a button to a user just to tell him he cannot perform the action if he tries), but the primarily goal is to prevent users from creating inconsistent data by performing "legal" actions in the Admin UI.

This work is sponsored by SWITCH

doofy committed:
MH-12815 warn if deleting a series with events

- add hasEvents get method to series endpoint
- new modal confirmation dialog for delete single series
- delete modal dialog warns if series is not empty

doofy committed:
MH-12815 delete series with events option

- config file for series endpoint
- config option if deletion of non empty series should be allowed or not
- new options method for series endpoint

doofy committed:
MH-12815 add allow option for series with events to bulk delete

- add column in series bulk delete modal to show which series has events
- add check to bulk delete button depending on configuration

doofy committed:
MH-12815 hints for better configuration in new series endpoint config file

staubesv committed:
Improve usability of deleting series with events

lkiesow commented on etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg:
```diff
@@ -0,0 +1,3 @@
+# If false a user will not be able to delete a series containing events in Admin UI
+# Note: Have a look at org.opencastproject.ui.metadata.CatalogUIAdapterFactory-episode-common.cfg property.series.required=false
+series.hasEvents.delete.allow=true
```
Can we set this to the form

```
# Long description. Some text. More text.
# Default: true
#series.hasEvents.delete.allow=true
```

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -207,6 +214,18 @@ protected void activate(ComponentContext cc) {
+    }
+
+    Object dictionaryValue = properties.get(SERIES_HASEVENTS_DELETE_ALLOW_KEY);
+    deleteSeriesWithEventsAllowed = Boolean.parseBoolean(dictionaryValue.toString());
```
Can we use [`BooleanUtils.toBoolean(String)`](https://commons.apache.org/proper/commons-lang/javadocs/api-3.0/org/apache/commons/lang3/BooleanUtils.html#toBoolean(java.lang.String)) since it is more robust?

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -1021,6 +1040,35 @@ private boolean hasProcessingEvents(String seriesId) {
+      SearchResult<Event> result = searchIndex.getByQuery(query);
+      elementsCount = result.getHitCount();
+    } catch (SearchIndexException e) {
+      logger.warn("Could not perform search query: {}", ExceptionUtils.getStackTrace(e));
```
Never use `ExceptionUtils` in Opencast. This will do nicely:

```java
logger.warn("Could not perform search query", e);
```

lkiesow commented on etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg:
```diff
@@ -0,0 +1,3 @@
+# If false a user will not be able to delete a series containing events in Admin UI
+# Note: Have a look at org.opencastproject.ui.metadata.CatalogUIAdapterFactory-episode-common.cfg property.series.required=false
```
Why should I take a look at that configuration option?

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -1039,4 +1087,16 @@ private boolean hasProcessingEvents(String seriesId) {
   }
 
+  @GET
+  @Path("options.json")
```
I was a bit confused by this name and did not expect this to return the service configuration. Maybe call this `/configuration.json`?

lkiesow commented on modules/admin-ui/src/main/resources/public/org/opencastproject/adminui/languages/lang-en_US.json:
```diff
@@ -64,7 +64,13 @@
+       "CONFIRMATION" : "Confirmation"
+     },
+     "WARNINGS": {
+       "SERIES_HAS_EVENTS": "This series does contain events. Deleting the series will not delete the events"
```
`This series does contain events. Deleting the series will not delete the events*.*`

If you use a full stop, there should also be one at the end.

lkiesow commented on modules/admin-ui/src/main/resources/public/org/opencastproject/adminui/languages/lang-en_US.json:
```diff
@@ -64,7 +64,13 @@
+       "SERIES_HAS_EVENTS": "This series does contain events. Deleting the series will not delete the events"
+     },
+     "ERRORS": {
+       "SERIES_HAS_EVENTS": "This series cannot be deleted as it still contains events"
```
`.` here as well?

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/modules/events/controllers/bulkDeleteController.js:
```diff
@@ -107,6 +108,18 @@ angular.module('adminNg.controllers')
 
+    $scope.allowed = function () {
+        var allowed = true;
+        if (Table.resource.indexOf('series') >= 0 && $scope.deleteSeriesWithEventsAllowed == false) {
```
`if (Table.resource.indexOf('series') >= 0 && !$scope.deleteSeriesWithEventsAllowed) {`
￼

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/modules/events/controllers/bulkDeleteController.js:
```diff
@@ -201,4 +214,16 @@ angular.module('adminNg.controllers')
+        SeriesOptionsResource.get(function (data) {
+            $scope.deleteSeriesWithEventsAllowed = data.deleteSeriesWithEventsAllowed;
+        });
+        var result = [];
```
This is never used, is it?

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/shared/partials/modals/delete-single-series-modal.html:
```diff
@@ -0,0 +1,44 @@
+        <h2 translate="CONFIRMATIONS.ACTIONS.CONFIRMATION"><!-- Confirmation  --></h2>
+    </header>
+
+    <div ng-if="hasEvents == true && deleteSeriesWithEventsAllowed == false">
```
`<div ng-if="hasEvents && !deleteSeriesWithEventsAllowed">￼`?

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/shared/partials/modals/delete-single-series-modal.html:
```diff
@@ -0,0 +1,44 @@
+      </div>
+    </div>
+
+    <div ng-if="hasEvents == false || deleteSeriesWithEventsAllowed == true">
```
`<div ng-if="!hasEvents || deleteSeriesWithEventsAllowed">`?

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/shared/partials/modals/delete-single-series-modal.html:
```diff
@@ -0,0 +1,44 @@
+
+    <div ng-if="hasEvents == false || deleteSeriesWithEventsAllowed == true">
+
+      <div ng-if="hasEvents == true && deleteSeriesWithEventsAllowed == true" class="modal-alert warning">
```
Again: Why compare to boolean value?

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/shared/services/deleteSingleSeriesModalService.js:
```diff
@@ -0,0 +1,89 @@
+            //64 picked by random experimentation
+            if (me.$scope.name.length > 64) {
+                me.$scope.name = me.$scope.name.substr(0,61);
+                me.$scope.name = me.$scope.name + "...";
```
Use ellipsis Unicode character: `…`

lkiesow commented on modules/admin-ui/src/main/webapp/scripts/shared/services/deleteSingleSeriesModalService.js:
```diff
@@ -0,0 +1,89 @@
+            if (!angular.isUndefined(object)) {
+                me.$scope.id = object.id;
+                if (object.title) {
+                    me.$scope.name = object.title;
```
Why switch from `title` to `name`?

doofy replied:
Made sense to me at that point in time. I fear the overhead changing this now.

lkiesow commented on modules/admin-ui/src/main/webapp/styles/views/modals/_modal-dialog.scss:
```diff
@@ -70,4 +70,41 @@
+
+        p {
+            text-align: center;
+            margin:20px auto 20px;
```
`margin: 20px auto;`

doofy replied:
besides the missing leading space this is intentional

lkiesow reviewed (commented):
Looks good. Tested this and it works as expected. I have added a few comments about minor issues though.

It would also be great if you could add a very short feature notice to the [release docs](https://github.com/opencast/opencast/blob/develop/docs/guides/admin/docs/releasenotes.md#new-features-and-improvements)

doofy committed:
MH-12815 small notice in release docs

doofy committed:
MH-12815 better config file option handling

- follow best practice for config file commenting
- use default when config key is not defined

doofy committed:
MH-12815 rename endpoint to configuration

doofy committed:
MH-12815 implement reviewer notes

- use BooleanUtils insted of Boolean class to parse boolean
- do not use exceptionutils
- consistent stops in translation text
- simplify boolean logic statements
- remove unused code
- use correct symbol for ...
- fix css

doofy commented:
Dear @lkiesow, I fixed almost all your notes. Please continue with your review process and thanks!

doofy commented:
I think the failing test is not correlated with the changes in this PR. Please see #364

doofy committed:
MH-12815 consistent syntax to please markdown lint

lkiesow commented:
Thanks for the patch and fixing the issues → merged
